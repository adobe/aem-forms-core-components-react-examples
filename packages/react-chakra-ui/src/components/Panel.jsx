import React, { useContext } from "react";
import { SimpleGrid, Text } from "@chakra-ui/react";

import {
  FormContext,
  useRuleEngine,
  renderChildren,
  useFormIntl
} from "@aemforms/af-react-renderer";
import { getOrElse } from '@aemforms/af-core';

const Panel = (fieldset) => {
  const context = useContext(FormContext);
  const [props, handlers] = useRuleEngine(fieldset);

  const {
    layout
  } = props?.properties || {};
  const { columns = "", className = "" } = layout || {};

  const translationId = getOrElse(props, ['properties', 'afs:translationIds', 'label.value']);

  const i18n = useFormIntl();
  let localizedLabel = props?.label?.value;
  if (translationId) {
    localizedLabel = i18n.formatMessage({ id: translationId, defaultMessage: props?.label?.value });
  }

  if (props.visible) {
    return (
      <>
      {props.label?.visible !== false ? <Text>{localizedLabel}</Text> : null}
        <SimpleGrid
        columnGap={10}
        columns={columns}
        colorScheme="blue"
        className={className}
      >
        {renderChildren(props, context.mappings, context.modelId, handlers)}
      </SimpleGrid>
      </>

    );
  } else {
    return null;
  }
};

export default Panel;
