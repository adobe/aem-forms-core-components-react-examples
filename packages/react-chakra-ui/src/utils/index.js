import React from 'react';
import { FormContext } from '@aemforms/af-react-renderer';
import { createFormInstance } from '@aemforms/af-core';
import { ChakraProvider } from '@chakra-ui/react'
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';

export const DEFAULT_ERROR_MESSAGE = 'There is an error in the field';

export const createForm = (field) => {
  const formJson = {
    items: [field]
  };
  return createFormInstance(formJson);
};

export const Provider = (form, mappings, locale, dictionaries) => props => {
  const c = {
    form,
    mappings,
    modelId: '$form',
    refMap: {}
  };
  const { children } = props;
  return (
    <ChakraProvider>
      <IntlProvider locale={locale} messages={dictionaries}>
        <FormContext.Provider value={c}>{children}</FormContext.Provider>
      </IntlProvider>
    </ChakraProvider>
  );
};

export const renderComponent = function (Component) {
  const test = (field, operation) => {
    const form = createForm(field);
    if (operation) {
      operation(form, form.items[0]);
    }
    const e = form.items[0].getState();
    let component = <Component {...e} />;
    const wrapper = Provider(form);
    const renderResponse = render(component, { wrapper });
    return {
      renderResponse,
      form,
      element: form?.items[0],
      component
    };
  };
  return test;
};
