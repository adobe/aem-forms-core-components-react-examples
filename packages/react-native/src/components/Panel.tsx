import { FieldsetJson, State, getOrElse } from '@aemforms/af-core';
import React, { useContext } from 'react';
import { useRuleEngine, useFormIntl, renderChildren, FormContext } from '@aemforms/af-react-renderer';
import { Box, Text } from 'native-base';

const Panel = function (fieldset: State<FieldsetJson>) {
  const context = useContext(FormContext);
  const [props, handlers] = useRuleEngine(fieldset);
  const translationId = getOrElse(props, ['properties', 'afs:translationIds', 'label.value']);
  const i18n = useFormIntl();
  let localizedLabel = props?.label?.value;
  if (translationId) {
    localizedLabel = i18n.formatMessage({ id: translationId, defaultMessage: props?.label?.value });
  }

  if (props.visible) {
    return (
      <Box>
        <Text>{localizedLabel}</Text>
        {renderChildren(props, context.mappings, context.modelId, handlers)}
      </Box>);
  } else {
    return null;
  }
};

export default Panel;