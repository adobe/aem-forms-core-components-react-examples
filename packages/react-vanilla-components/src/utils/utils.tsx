export const formatBytes = (bytes: number, decimals = 0) => {
    if (!+bytes) {
      return '0 Bytes';
    }
    const k = 1024;
    const dm = decimals < 0 ? 0 : decimals;
    const sizes = ['bytes', 'kb', 'mb', 'gb', 'tb', 'pb'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));

    return `${parseFloat((bytes / Math.pow(k, i)).toFixed(dm))} ${sizes[i]}`;
  };

  export const syncAriaDescribedBy = (id: string, tooltip?: string, description?: string, error?: string) => {
    const descriptions = [];

    if (description) {
        descriptions.push(`${id}__longdescription`);
    }
    if (tooltip) {
        descriptions.push(`${id}__shortdescription`);
    }
    if (error) {
        descriptions.push(`${id}__errormessage`);
    }

    return descriptions.join(' ');
};