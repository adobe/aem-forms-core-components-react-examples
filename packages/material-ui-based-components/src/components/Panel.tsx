import React, { useContext } from 'react';
import { renderChildren, FormContext } from '@aemforms/af-react-renderer';
import { withRuleEnginePanel } from '../shared/withRuleEngine';
import { PROPS_PANEL } from '../utils/types';

const Panel = function (props: PROPS_PANEL) {
  const { handlers, ...restProps } = props;
  // @ts-ignore
  const context = useContext(FormContext);
  return (
    <div>
      {restProps?.label?.visible ? <div>{restProps.label.value}</div> : null}
      {renderChildren(restProps, context.mappings, context.modelId, handlers)}
    </div>
  );
};

export default withRuleEnginePanel(Panel);