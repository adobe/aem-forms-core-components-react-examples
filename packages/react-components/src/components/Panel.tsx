import { State, FieldsetJson } from '@aemforms/af-core';
import React, { useContext, useState } from 'react';
import withRuleEngine from '../utils/HOC';
import { IHandler } from '../utils/type';
import { renderChildren, FormContext } from '@aemforms/af-react-renderer';
import '../../dist/theme.css';

const Panel = function (props: State<FieldsetJson & IHandler>) {
  const {id,visible,enabled,label,description} = props;
  const context = useContext(FormContext);
  const {dispatchAddItem, dispatchBlur, dispatchChange, dispatchClick, dispatchFocus, dispatchRemoveItem} = props;
  const handlers: IHandler = {dispatchAddItem, dispatchBlur, dispatchChange, dispatchClick, dispatchFocus, dispatchRemoveItem};
  const [show, setShow] = useState(true);
  const handleClick = () => {
    setShow(!show);
  };
  console.log('panel-props', props);
    return (
      <div id={id} data-cmp-visible={visible} data-cmp-enabled={enabled} data-cmp-is="adaptiveFormPanel">
        {label?.visible && <label htmlFor={id} className="cmp-container__label">{label?.value}</label>}
        <button className="cmp-container__questionmark" onClick={handleClick}></button>
        <div aria-live="polite">
            {show && description && <div className='cmp-container__longdescription' data-cmp-visible={show}>{description}</div>}
        </div>
        {renderChildren(props, context.mappings, context.modelId, handlers)}
      </div>
    );
};

export default withRuleEngine(Panel);