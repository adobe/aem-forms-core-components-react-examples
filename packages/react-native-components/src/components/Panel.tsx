/*************************************************************************
* ADOBE CONFIDENTIAL
* ___________________
*
* Copyright 2023 Adobe
* All Rights Reserved.
*
* NOTICE: All information contained herein is, and remains
* the property of Adobe and its suppliers, if any. The intellectual
* and technical concepts contained herein are proprietary to Adobe
* and its suppliers and are protected by all applicable intellectual
* property laws, including trade secret and copyright laws.
* Dissemination of this information or reproduction of this material
* is strictly forbidden unless prior written permission is obtained
* from Adobe.

* Adobe permits you to use and modify this file solely in accordance with
* the terms of the Adobe license agreement accompanying it.
*************************************************************************/

import React, { useContext } from 'react';
import { renderChildren, FormContext } from '@aemforms/af-react-renderer';
import { Box, Text } from 'native-base';
import { withRuleEnginePanel } from '../shared/withRuleEngine';
import { PROPS_PANEL } from '../utils/types';

const Panel = function (props: PROPS_PANEL) {
  const { label, handlers } = props;
  const context = useContext(FormContext);

  return (
    <Box>
      {label?.visible !== false ? <Text>{label?.value}</Text> : null}
      {renderChildren(props, context.mappings, context.modelId, handlers)}
    </Box>
  );
};

export default withRuleEnginePanel(Panel);