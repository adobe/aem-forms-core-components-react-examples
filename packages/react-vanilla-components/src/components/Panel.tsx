// *******************************************************************************
//  * Copyright 2023 Adobe
//  *
//  * Licensed under the Apache License, Version 2.0 (the “License”);
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *     http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an “AS IS” BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.

//  * The BEM markup is as per the AEM core form components guidelines.
//  * LINK- https://github.com/adobe/aem-core-forms-components/blob/master/ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/panelcontainer/v1/panelcontainer/panelcontainer.html
//  ******************************************************************************

import React, { useContext } from 'react';
import { withRuleEnginePanel } from '../utils/withRuleEngine';
import { renderChildren, FormContext } from '@aemforms/af-react-renderer';
import { PROPS_PANEL } from '../utils/type';
import LabelWithDescription from './common/LabelWithDescription';

const Panel = function (props: PROPS_PANEL) {
  const { id, visible, enabled, label, handlers, appliedCssClassNames } = props;
  // @ts-ignore
  const context = useContext(FormContext);

  return (
    <div
      className={`cmp-container ${appliedCssClassNames || ''}`}
      id={id}
      data-cmp-visible={visible}
      data-cmp-enabled={enabled}
      data-cmp-is="adaptiveFormPanel"
    >
      <LabelWithDescription
        bemBlock='cmp-container'
        label={label}
        id={id}
        tooltip={props.tooltip}
        description={props.description}
      />
      {renderChildren(props, context.mappings, context.modelId, handlers)}
    </div>
  );
};

export default withRuleEnginePanel(Panel);