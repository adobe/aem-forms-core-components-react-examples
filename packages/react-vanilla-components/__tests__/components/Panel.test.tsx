/*
 * Copyright 2023 Adobe, Inc.
 *
 * Your access and use of this software is governed by the Adobe Customer Feedback Program Terms and Conditions or other Beta License Agreement signed by your employer and Adobe, Inc.. This software is NOT open source and may not be used without one of the foregoing licenses. Even with a foregoing license, your access and use of this file is limited to the earlier of (a) 180 days, (b) general availability of the product(s) which utilize this software (i.e. AEM Forms), (c) January 1, 2023, (d) Adobe providing notice to you that you may no longer use the software or that your beta trial has otherwise ended.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL ADOBE NOR ITS THIRD PARTY PROVIDERS AND PARTNERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import Panel from '../../src/components/Panel';
import React from 'react';
import { render } from '@testing-library/react';
import { createForm, Provider, renderComponent } from '../utils';
import { jsonString } from '@aemforms/af-core';

const emptyPanel = {
  'id': 'emptypanel',
  'visible': true,
  'items': [],
  'label': {
    'value': 'panel title'
  },
  index: 0,
  ':type': 'core/fd/components/form/panelcontainer/v1/panelcontainer',
  "properties": {
    "afs:translationIds": {
      "label.value": "firstName##7081##title##1335",
      "description": "firstName##7081##description##9756"
    }
  },
};

const panelWithField = {
  'id': 'panelWithField',
  'visible': true,
  'label': {
    'value': 'panel title'
  },
  "properties": {
    "afs:translationIds": {
      "label.value": "firstName##7081##title##1335",
      "description": "firstName##7081##description##9756"
    }
  },
  index: 0,
  ':type': 'core/fd/components/form/panelcontainer/v1/panelcontainer',
  'items': [
    {
      'id': 'field',
      'name': 'f1',
      'fieldType': 'text-field',
      ':type': 'text-field',
      'title': 'name',
      'visible': true,
      index: 0
    }
  ]
};

test("Snapshot", () => {
  const helper = renderComponent(Panel);
  const { renderResponse } = helper(panelWithField);
  expect(renderResponse.container).toMatchSnapshot();
});

test('panel should get rendered if it is visible', () => {
  const helper = renderComponent(Panel);
  const { renderResponse } = helper(emptyPanel);
  expect(renderResponse.container.innerHTML.length).toBeGreaterThan(0);
});

test('panel should get rendered if it has no children', () => {
  const helper = renderComponent(Panel);
  const { renderResponse } = helper(emptyPanel);
  expect(renderResponse.container.innerHTML.length).toBeGreaterThan(0);
});


test('panel should get rendered if no mapping is defined', async () => {
  const panel = <Panel {...panelWithField} />;
  const form = await createForm(emptyPanel);
  const wrapper = Provider(form, {});
  const { container } = render(panel, { wrapper });
  expect(container.innerHTML.length).toBeGreaterThan(0);
});

test('children with mapping should render the mapped component', async () => {
  const MyComponent = () => {
    return <div>My Field</div>;
  };

  const panel = <Panel {...panelWithField} />;
  const form = await createForm(emptyPanel);
  const wrapper = Provider(form, { 'text-field': MyComponent });
  const { container } = render(panel, { wrapper });
  var expected = '<div><h4>Undefined Element</h4><pre>' + jsonString(panelWithField.items[0]) + '</pre></div>';
  expect(container.innerHTML).not.toContain(expected);
  expect(container.innerHTML).toContain('<div>My Field</div>');
});

test('panel should get rendered with label', async () => {
  const helper = renderComponent(Panel);
  const { renderResponse } = helper(panelWithField);
  expect(renderResponse.queryByText(panelWithField.label.value)).not.toBeNull();
});

test('html in the label should be handled for non rich text', async () => {
  const helper = renderComponent(Panel);
  const f = {
    ...panelWithField,
    label: {
      value: '<p>title inside p tags</p>',
      richText: true,
      visible: true
    }
  }
  let { renderResponse } = await helper(f);
  expect(renderResponse.container.innerHTML).toContain('<p>title inside p tags</p>');
});



