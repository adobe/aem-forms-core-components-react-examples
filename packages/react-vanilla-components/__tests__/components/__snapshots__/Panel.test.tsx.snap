// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
<div>
  <div
    class="cmp-container "
    data-cmp-enabled="true"
    data-cmp-is="adaptiveFormPanel"
    data-cmp-visible="true"
    id="panelWithField"
  >
    <div
      class="cmp-container__label-container"
    >
      <label
        class="cmp-container__label"
        for="panelWithField-widget"
      >
        panel title
      </label>
      <button
        aria-label="[object Object]"
        class="cmp-container__questionmark"
      />
    </div>
    <div
      aria-live="polite"
    />
  </div>
</div>
`;
